test/lit/recursive-types.wast

```diff
@@ -2,20 +2,12 @@
 
 ;; RUN: wasm-opt %s -all -S -o - | filecheck %s
 
-;; TODO: Fix the bug where self-referential function signatures are emitted
-;; twice. This happens because collectHeapTypes has to reconstruct a HeapType
-;; from each function's signature, but self-referential HeapTypes aren't
-;; canonicalized when they are parsed so collectHeapTypes ends up with a
-;; separate, unfolded version of the type.
-
 ;; TODO: Fix the bug where structurally identical types are given the same
 ;; generated name, making the wast invalid due to duplicate names.
 
 ;; CHECK:      (module
 ;; CHECK-NEXT:   (type $ref?|...0|_=>_ref?|...0| (func (param (ref null $ref?|...0|_=>_ref?|...0|)) (result (ref null $ref?|...0|_=>_ref?|...0|))))
 ;; CHECK-NEXT:   (type $ref?|...0|_=>_ref?|...0| (func (param (ref null $ref?|...0|_=>_ref?|...0|)) (result (ref null $ref?|...0|_=>_ref?|...0|))))
```

Comment:
Unrelated to this PR, this output looks like it emits two types with the exact same name, which should be invalid - that is, I think this file would error on round-tripping through the text format?

(I guess #3627 could fix that?)

Reply:
Yep, correct on both counts 👍